Sources/SwiftProtobuf/Message.swift

```diff
@@ -18,64 +18,100 @@
+/// `Message` is the protocol type you should use whenever
+/// you need an argument or variable which holds "some message".
 ///
-/// See ProtobufBinaryTypes and ProtobufJSONTypes for extensions
-/// to these protocols for supporting binary and JSON coding.
+/// Generated messages also implement `Hashable`, and thus `Equatable`.
+/// However, the protocol conformance is declared on a different protocol.
+/// This allows you to use `Message` as a type directly:
 ///
-
+///     func consume(message: Message) { ... }
 ///
-/// ProtobufMessage is the protocol type you should use whenever
-/// you need an argument or variable which holds "some message".
+/// Instead of needing to use it as a type constraint on a generic declaration:
 ///
-/// In particular, this has no associated types or self references so can be
-/// used as a variable or argument type.
+///     func consume<M: Message>(message: M) { ... }
 ///
+/// If you need to convince the compiler that your message is `Hashable` so
+/// you can insert it into a `Set` or use it as a `Dictionary` key, use
+/// a generic declaration with a type constraint:
+///
+///     func insertIntoSet<M: Message & Hashable>(message: M) {
+///         mySet.insert(message)
+///     }
+///
+/// The actual functionality is implemented either in the generated code or in
+/// default implementations of the below methods and properties. Some of them,
+/// including `hashValue` and `debugDescription`, are designed to let you
+/// override the functionality in custom extensions to the generated code.
 public protocol Message: CustomDebugStringConvertible {
+  /// Builds an instance of the message with all fields initialized to
+  /// their default values.
   init()
 
   // Metadata
   // Basic facts about this class and the proto message it was generated from
   // Used by various encoders and decoders
+  /// Name of this message's concrete Swift type
   var swiftClassName: String { get }
+
+  /// Name of the message from the original .proto file
   var protoMessageName: String { get }
+
+  /// Name of the protobuf package from the original .proto file
   var protoPackageName: String { get }
+
+  /// Prefix used for this message's type when encoded as an `Any`
   var anyTypePrefix: String { get }
+
+  /// Fully qualifed name used for this message's type when encoed as an `Any`
   var anyTypeURL: String { get }
 
   //
   // General serialization/deserialization machinery
   //
 
   /// Decode a field identified by a field number (as given in the .proto file).
-  /// The Message will call the FieldDecoder method corresponding
+  /// The `Message` will call the `FieldDecoder` method corresponding
   /// to the declared type of the field.
   ///
   /// This is the core method used by the deserialization machinery.
   ///
   /// Note that this is not specific to protobuf encoding; formats that use
   /// textual identifiers translate those to protoFieldNumbers and then invoke
   /// this to decode the field value.
```

Comment:
I can envision some people wanting to override this in order to avoid decoding certain fields.  To do so, the easiest way would be to leverage the `_protoc_generated_decodeField()` created by the code generator:
```
public mutating func decodeField<T: FieldDecoder>(setter: inout T, protoFieldNumber: Int) throws {
   switch protoFieldNumber {
   case 1: break // Don't decode this field at all
   default: try _protoc_generated_decodeField(setter: &setter, protoFieldNumber: protoFieldNumber)
   }
}
```